File: BlinkID/android/src/main/kotlin/com/microblink/blinkid/flutter/SerializationUtils.java
package com.microblink.blinkid.flutter;

import android.graphics.Bitmap;
import android.util.Base64;

import com.microblink.blinkid.entities.recognizers.Recognizer;
//import com.microblink.blinkid.entities.parsers.Parser;
import com.microblink.blinkid.geometry.Point;
import com.microblink.blinkid.geometry.Quadrilateral;
import com.microblink.blinkid.geometry.Rectangle;
import com.microblink.blinkid.image.Image;
import com.microblink.blinkid.results.date.SimpleDate;
import com.microblink.blinkid.results.date.Date;
import com.microblink.blinkid.entities.Entity;
import com.microblink.blinkid.entities.recognizers.blinkid.imageoptions.extension.ImageExtensionFactors;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

import androidx.annotation.Nullable;

public abstract class SerializationUtils {
    private static final int COMPRESSED_IMAGE_QUALITY = 90;

    public static <T extends Recognizer.Result> void addCommonRecognizerResultData(JSONObject jsonObject, T result) throws JSONException {
        jsonObject.put("resultState", serializeEnum(result.getResultState()));
    }

    public static JSONObject serializeSimpleDate( @Nullable  SimpleDate date ) throws JSONException {
        if (date != null ) {
            JSONObject jsonDate = new JSONObject();
            jsonDate.put("day", date.getDay());
            jsonDate.put("month", date.getMonth());
            jsonDate.put("year", date.getYear());
            return jsonDate;
        } else {
            return null;
        }
    }

    public static JSONObject serializeDate(@Nullable Date dateResult) throws JSONException {
        if (dateResult == null) {
            return null;
        } else {
            return serializeSimpleDate(dateResult.getDate());
        }
    }

    public static int serializeEnum(Enum e) {
        return e.ordinal();
    }

    public static JSONArray serializeStringArray(String[] strings) {
        JSONArray jsonStrings = new JSONArray();
        for (String str : strings) {
            jsonStrings.put(str);
        }
        return jsonStrings;
    }

    public static String encodeImageBase64(Image image) {
        if (image == null) {
            return null;
        }
        Bitmap resultImgBmp = image.convertToBitmap();
        if (resultImgBmp == null) {
            return null;
        }
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        boolean success = resultImgBmp.compress(Bitmap.CompressFormat.JPEG, COMPRESSED_IMAGE_QUALITY, byteArrayOutputStream);
        String resultImgBase64 = null;
        if (success) {
            resultImgBase64 = Base64.encodeToString(byteArrayOutputStream.toByteArray(), Base64.NO_WRAP);
        }
        try {
            byteArrayOutputStream.close();
        } catch (IOException ignorable) {}
        return resultImgBase64;
    }

    public static String encodeByteArrayToBase64(byte[] arr) {
        if (arr == null) {
            return null;
        }
        return Base64.encodeToString(arr, Base64.NO_WRAP);
    }

    public static JSONObject serializePoint(Point point) throws JSONException {
        JSONObject jsonPoint = new JSONObject();
        jsonPoint.put("x", point.getX());
        jsonPoint.put("y", point.getY());
        return jsonPoint;
    }

    public static JSONObject serializeQuad(Quadrilateral quad) throws JSONException {
        JSONObject jsonQuad = new JSONObject();
        jsonQuad.put("upperLeft", serializePoint(quad.getUpperLeft()));
        jsonQuad.put("upperRight", serializePoint(quad.getUpperRight()));
        jsonQuad.put("lowerLeft", serializePoint(quad.getLowerLeft()));
        jsonQuad.put("lowerRight", serializePoint(quad.getLowerRight()));
        return jsonQuad;
    }

    public static JSONObject serializeRectangle(Rectangle rectangle) throws JSONException {
        JSONObject jsonRectangle = new JSONObject();
        if (rectangle != null) {
            jsonRectangle.put("x", rectangle.getX());
            jsonRectangle.put("y", rectangle.getY());
            jsonRectangle.put("height", rectangle.getHeight());
            jsonRectangle.put("width", rectangle.getWidth());
        }
        return jsonRectangle;
    }

    public static String getStringFromJSONObject(JSONObject map, String key) {
        String value = map.optString(key, null);
        if ("null".equals(value)) {
            value = null;
        }
        return value;
    }

    public static ImageExtensionFactors deserializeExtensionFactors(JSONObject jsonExtensionFactors) {
        if (jsonExtensionFactors == null) {
            return new ImageExtensionFactors(0.f, 0.f, 0.f, 0.f);
        } else {
            float up = (float)jsonExtensionFactors.optDouble("upFactor", 0.0);
            float right = (float)jsonExtensionFactors.optDouble("rightFactor", 0.0);
            float down = (float)jsonExtensionFactors.optDouble("downFactor", 0.0);
            float left = (float)jsonExtensionFactors.optDouble("leftFactor", 0.0);
            return new ImageExtensionFactors(up, down, left, right);
        }
    }

}

File: BlinkID/android/src/main/kotlin/com/microblink/blinkid/flutter/SerializationUtils.kt
package com.microblink.blinkid.flutter

import android.graphics.Bitmap
import android.util.Base64
import com.microblink.blinkid.entities.recognizers.Recognizer
import com.microblink.blinkid.entities.recognizers.blinkid.imageoptions.extension.ImageExtensionFactors
import com.microblink.blinkid.geometry.Point
import com.microblink.blinkid.geometry.Quadrilateral
import com.microblink.blinkid.image.Image
import com.microblink.blinkid.results.date.Date
import com.microblink.blinkid.results.date.SimpleDate
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.IOException

//import com.microblink.blinkid.entities.parsers.Parser;
object SerializationUtils {
    private const val COMPRESSED_IMAGE_QUALITY = 90
    @JvmStatic
    @Throws(JSONException::class)
    fun <T : Recognizer.Result?> addCommonRecognizerResultData(jsonObject: JSONObject, result: T) {
        jsonObject.put("resultState", serializeEnum(result!!.resultState))
    }

    /*public static <T extends Parser.Result> void addCommonParserResultData(JSONObject jsonObject, T result) throws JSONException {
        jsonObject.put("resultState", serializeEnum(result.getResultState()));
    }*/
    @JvmStatic
    @Throws(JSONException::class)
    fun serializeSimpleDate(date: SimpleDate?): JSONObject? {
        return if (date != null) {
            val jsonDate = JSONObject()
            jsonDate.put("day", date.day)
            jsonDate.put("month", date.month)
            jsonDate.put("year", date.year)
            jsonDate
        } else {
            null
        }
    }

    @JvmStatic
    @Throws(JSONException::class)
    fun serializeDate(dateResult: Date?): JSONObject? {
        return if (dateResult == null) {
            null
        } else {
            serializeSimpleDate(dateResult.date)
        }
    }

    @JvmStatic
    fun serializeEnum(e: Enum<*>): Int {
        return e.ordinal
    }

    @JvmStatic
    fun serializeStringArray(strings: Array<String?>): JSONArray {
        val jsonStrings = JSONArray()
        for (str in strings) {
            jsonStrings.put(str)
        }
        return jsonStrings
    }

    @JvmStatic
    fun encodeImageBase64(image: Image?): String? {
        if (image == null) {
            return null
        }
        val resultImgBmp = image.convertToBitmap() ?: return null
        val byteArrayOutputStream = ByteArrayOutputStream()
        val success = resultImgBmp.compress(Bitmap.CompressFormat.JPEG, COMPRESSED_IMAGE_QUALITY, byteArrayOutputStream)
        var resultImgBase64: String? = null
        if (success) {
            resultImgBase64 = Base64.encodeToString(byteArrayOutputStream.toByteArray(), Base64.NO_WRAP)
        }
        try {
            byteArrayOutputStream.close()
        } catch (ignorable: IOException) {
        }
        return resultImgBase64
    }

    @JvmStatic
    fun encodeByteArrayToBase64(arr: ByteArray?): String? {
        return if (arr == null) {
            null
        } else Base64.encodeToString(arr, Base64.NO_WRAP)
    }

    @Throws(JSONException::class)
    fun serializePoint(point: Point): JSONObject {
        val jsonPoint = JSONObject()
        jsonPoint.put("x", point.x.toDouble())
        jsonPoint.put("y", point.y.toDouble())
        return jsonPoint
    }

    @JvmStatic
    @Throws(JSONException::class)
    fun serializeQuad(quad: Quadrilateral): JSONObject {
        val jsonQuad = JSONObject()
        jsonQuad.put("upperLeft", serializePoint(quad.upperLeft))
        jsonQuad.put("upperRight", serializePoint(quad.upperRight))
        jsonQuad.put("lowerLeft", serializePoint(quad.lowerLeft))
        jsonQuad.put("lowerRight", serializePoint(quad.lowerRight))
        return jsonQuad
    }

    @JvmStatic
    fun getStringFromJSONObject(map: JSONObject, key: String?): String? {
        var value = map.optString(key, null)
        if ("null" == value) {
            value = null
        }
        return value
    }

    @JvmStatic
    fun deserializeExtensionFactors(jsonExtensionFactors: JSONObject?): ImageExtensionFactors {
        return if (jsonExtensionFactors == null) {
            ImageExtensionFactors(0f, 0f, 0f, 0f)
        } else {
            val up = jsonExtensionFactors.optDouble("upFactor", 0.0).toFloat()
            val right = jsonExtensionFactors.optDouble("rightFactor", 0.0).toFloat()
            val down = jsonExtensionFactors.optDouble("downFactor", 0.0).toFloat()
            val left = jsonExtensionFactors.optDouble("leftFactor", 0.0).toFloat()
            ImageExtensionFactors(up, down, left, right)
        }
    }
}